README.md
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/sensu/sensu-go-fatigue-check-filter)
![Karma Test](https://github.com/sensu/sensu-go-fatigue-check-filter/workflows/Karma%20Test/badge.svg)
![Otto Test](https://github.com/sensu/sensu-go-fatigue-check-filter/workflows/Otto%20Test/badge.svg)
![release](https://github.com/sensu/sensu-go-fatigue-check-filter/workflows/release/badge.svg)

## Sensu Go Fatigue Check Filter

- [Overview](#overview)
- [Usage examples](#usage-examples)
- [Configuration](#configuration)
  - [Sensu Go](#sensu-go)
    - [Asset registration](#asset-registration)
    - [Asset definition](#asset-definition)
    - [Filter definition](#filter-definition)
    - [Handler definition](#handler-definition)
    - [Check definition](#check-definition)
    - [Entity definition](#entity-definition)
    - [Keepalives](#keepalives)
    - [Cron scheduled checks](#cron-scheduled-checks)
    - [Annotations](#annotations)
    - [Arguments](#arguments)
    - [Non-repeating alerts](#non-repeating-alerts)
- [Installation from source](#installation-from-source)
- [Additional notes](#additional-notes)
- [Contributing](#contributing)

### Overview

The Sensu Go Fatigue Check Filter is a [Sensu Event Filter][1] for managing
alert fatigue.

A typical use of filters is to reduce [alert fatigue][2].  One of the most
typical examples of this is create the following filter that only passes
through events on their first occurrence and every hour after that.

```yml
---
type: EventFilter
api_version: core/v2
metadata:
  name: hourly
  namespace: default
spec:
  action: allow
  expressions:
  - event.check.occurrences == 1 || event.check.occurrences % (3600 / event.check.interval)
    == 0
  runtime_assets: []
```

However, the use of the filter above creates some limitations.  Suppose you
have one check in particular that you want to change to only alert after three
(3) occurrences.  Typically that might mean creating another handler and filter
pair to assign to that check.  If you have to do this often enough and you
start to have an unwieldy mass of handlers and filters.

That's where this Fatigue Check Filter comes in.  Using annotations, it makes
the number of occurrences and the interval tunable on a per-check or per-entity
basis.  It also allows you to control whether or not resolution events are
passed through.

## Usage examples

N/A

## Configuration
### Sensu Go
#### Asset registration

Assets are the best way to make use of this plugin. If you're not using an
asset, please consider doing so! If you're using sensuctl 5.13 or later, you
can use the following command to add the asset: 

`sensuctl asset add sensu/sensu-go-fatigue-check-filter --rename fatigue-check-filter`

Note that the `--rename` is not necessary, but references to the runtime asset
in the filter definition as in the example below would need to be updated to
match. 

If you're using an earlier version of sensuctl, you can download the asset
definition from [this project's Bonsai asset index page][7].

You can create your own [asset][3] by creating a tar file containing
`lib/fatigue_check.js` and creating your asset definition accordingly.

#### Asset definition

If not using `sensuctl asset add`:

```yml
---
type: Asset
api_version: core/v2
metadata:
  name: fatigue-check-filter
  namespace: default
spec:
  sha512: 2e67975df7d993492cd5344edcb9eaa23b38c1eef7000576b396804fc2b33362b02a1ca2f7311651c175c257b37d8bcbbce1e18f6dca3ca04520e27fda552856
  url: http://example.com/sensu/assets/fatigue-check.tar.gz
```

#### Filter definition

```yml
---
type: EventFilter
api_version: core/v2
metadata:
  name: fatigue_check
  namespace: default
spec:
  action: allow
  expressions:
  - fatigue_check(event)
  runtime_assets:
  - fatigue-check-filter
```

#### Handler definition

```yml
---
type: Handler
api_version: core/v2
metadata:
  namespace: default
  name: email
spec:
  type: pipe
  command: sensu-email-handler -f from@example.com -t to@example.com -s smtp.example.com
    -u emailuser -p sup3rs3cr3t
  timeout: 10
  filters:
  - is_incident
  - not_silenced
  - fatigue_check
```

#### Check definition

```yml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: linux-cpu-check
  namespace: default
  annotations:
    fatigue_check/occurrences: '3'
    fatigue_check/interval: '900'
    fatigue_check/allow_resolution: 'false'
spec:
  command: check-cpu -w 90 c 95
  handlers:
  - email
  interval: 60
  publish: true
  runtime_assets: 
  subscriptions:
  - linux
```

#### Entity definition

Via the agent.yml:

```
---
##
# agent configuration
##

#name: ""

#namespace: "default"

#subscriptions: 
#  - "localhost"

annotations:
  fatigue_check/occurrences: "3"
  fatigue_check/interval: "900"
  fatigue_check/keepalive_occurrences: "1"
  fatigue_check/keepalive_interval: "300"
  fatigue_check/allow_resolution: "false"

[...]
```
#### Keepalives
Keepalives do not have check resources with annotations that can be used to tune
this filter.  Using standard entity annotations would override the settings for
**all** other checks.  To address this specific case, two additional tunables
exist for customizing this filter for keepalive events.  These can be set as
arguments to the `fatigue_check()` function in the filter definition or as
entity annotations to override the defaults on a per entity basis.

#### Cron scheduled checks
Since cron scheduled checks do not provide an explicit interval, this filter
has to compute the apparent interval using the history available in the event.
In order for this to work, there has to be at least two entries in the event's
check history.  The only time there would be less than two history entries is
during the first two check executions.  By default this should not be an issue,
but in the unlikely case the interval is set this low, a default interval of 60
seconds is used.

#### Annotations
The Fatigue Check Filter makes use of four annotations within the check and/or
entity metadata for normal checks with an additional two keepalive annotations
availalbe in the entity metadata.  The entity annotations taking precedence over
check annotations.  All annotations take precedence of the `fatigue_check()`
function arguments and defaults.

|Annotation|Default|Usage|
|----------|-------|-----|
|fatigue_check/occurrences|1|On which occurrence to allow the initial event to pass through for normal checks|
|fatigue_check/interval|1800|In seconds, at what interval to allow subsequent events to pass through, ideally a multiple of the check interval for normal checks|
|fatigue_check/allow_resolution|true|Determines whether or not a resolution event is passed through|
|fatigue_check/suppress_flapping|true|Determines whether or not to suppress events for checks that are marked as flapping|
|fatigue_check/keepalive_occurrences|1|On which occurrence to allow the initial event to pass through for keepalives (**entity only**)|
|fatigue_check/keepalive_interval|1800|In seconds, at what interval to allow subsequent events to pass through, ideally a multiple of the check interval for keepalives (**entity only**)|

#### Arguments
The `fatigue_check()` function can take up to five arguments.

```javascript
fatigue_check(event, occurrences, interval, keepalive_occurrences, keepalive_interval)
```

The first one is the event and is required.  The remaining four are optional and
allow you to override the built-in defaults for occurrences, interval,
keepalive_occurrences, and keepalive_interval, respectively.  For example, if
you'd like a version of the filter that, by default on non-keepalive checks,
matches on the second occurrence instead of the first you could create a filter
similar to below:

```yml
---
type: EventFilter
api_version: core/v2
metadata:
  name: fatigue_check_two_occurrences
  namespace: default
spec:
  action: allow
  expressions:
  - fatigue_check(event, 2)
  runtime_assets:
  - fatigue-check-filter
```

If you'd like one that overrides the default 30 minute interval for
non-keepalive checks with a 10 minute one you could create one similar to
below (note that in order to specify the third argument, you have to provide
the second):

```yml
---
type: EventFilter
api_version: core/v2
metadata:
  name: fatigue_check_10m_interval
  namespace: default
spec:
  action: allow
  expressions:
  - fatigue_check(event, 1, 600)
  runtime_assets:
  - fatigue-check-filter
```

If you'd like one that overrides the default occurrences for keeaplives
and alerts on the second occurrence rather than the first you could create one
similar to below (note that in order to specify the fourth argument, you have to
provide the second and third):

```yml
---
type: EventFilter
api_version: core/v2
metadata:
  name: fatigue_check_two_occurrences
  namespace: default
spec:
  action: allow
  expressions:
  - fatigue_check(event, 1, 1800, 2)
  runtime_assets:
  - fatigue-check-filter
```

If you'd like one that overrides the default 30 minute interval for
keepalives with a 10 minute one you could create one similar to below
(note that in order to specify the fifth argument, you have to provide
the second, third, and fourth as well, even if you want the defaults):

```yml
---
type: EventFilter
api_version: core/v2
metadata:
  name: fatigue_check_10m_interval
  namespace: default
spec:
  action: allow
  expressions:
  - fatigue_check(event, 1, 1800, 1, 600)
  runtime_assets:
  - fatigue-check-filter
```

#### Non-repeating alerts

If you need to have alerts which will not repeat, meaning the alert is only ever
sent on the first occurrence and none after (aside from the resolution, if
`allow_resolution` is true, which is the default), then you will need to set the
`interval` (or `keepalive_interval`) to zero (0) via an annotation.

## Installation from source

### Sensu Go

See the instructions above for [asset registration][6].

## Additional notes

* This filter makes use of the occurrences_watermark attribute that was buggy
up until Sensu Go 5.9. Your mileage may vary on prior versions.

* If the interval is not a multiple of the check's interval, then the actual
interval is computed by rounding up the result of dividing the interval by the
check's interval.  For example, an interval of 180s with a check interval of
25s would pass the event through on every 8 occurrences (200s).

## Contributing

Please submit an [issue][8] if you have problems or suggestions.

[1]: https://docs.sensu.io/sensu-go/latest/reference/filters/
[2]: https://docs.sensu.io/sensu-go/latest/guides/reduce-alert-fatigue/
[3]: https://docs.sensu.io/sensu-go/latest/reference/assets/
[4]: https://bonsai.sensu.io/
[5]: https://docs.sensu.io/sensu-core/latest/installation/installing-plugins/
[6]: #asset-registration
[7]: https://bonsai.sensu.io/assets/sensu/sensu-go-fatigue-check-filter
[8]: https://github.com/sensu/sensu-go-fatigue-check-filter/issues


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.8.1] - 2021-03-31

### Changed
- cleaned up README

## [0.8.0] - 2020-11-24

### Added
- Support for interval checking with cron scheduled checks

## [0.7.0] - 2020-07-27

### Added
- Special case handling for keepalive checks
- Test cases for zero interval

### Changed
- Reformatted and cleaned up README
- Added information on non-repeating alerts to README

## [0.6.2] - 2020-05-10

### Changed
- Updated parameter call for defaults, Otto doesn't support assigning default values in function definition

## [0.6.1] - 2020-05-08

### Changed
- Fixed releast to use all github actions, no included scripts

## [0.6.0] - 2020-05-08

### Changed
- Added default occurrences and interval as parameters to allow override(s)
- Reorganized README, added above and switched examples to yaml

## [0.5.0] - 2020-02-19

### Changed
- Minor github upload script changes
- Use bracket accessor syntax for annotations (@flowerysong)

## [0.4.0] - 2020-02-10

### Changed
- Migrated from Travis CI to GitHub Actions

## [0.3.2] - 2019-12-03

### Changed
- Reformatted README for [Plugin Style Guide](https://github.com/sensu-plugins/community/blob/master/PLUGIN_STYLEGUIDE.md)

### Added
- Testing with Karma
- Build and release scripts

### Removed
- Janky test script lib/fatigue_check_test.js

## [0.3.1] - 2019-12-03

### Changed
- Fixed is_resolution checking

## [0.3.0] - 2019-11-27

### Added
- Support for entity annotations

## [0.2.2] - 2019-09-21

### Changed
- README updated to include sensuctl asset add for bonsai

## [0.2.1] - 2019-09-04

### Changed
- README fixed for syntax error, forced version update to refresh bonsai

## [0.2.0] - 2019-08-30

### Changed
- Rounding the interval/check interval up just in case it's not a multiple

## [0.1.3] - 2019-06-18

### Changed
- More README fixes for clarity, forced version update to refresh bonsai

## [0.1.2] - 2019-06-18

### Changed
- Fixed README for Bonsai

## [0.1.0] - 2019-06-18

### Changed
- Updated to use occurrences_watermark

### Added
- Suppression of flapping events



karma.conf.js
// Karma configuration
module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],

    // list of files to exclude
    exclude: [
    ],

    // list of files / patterns to load in the browser
    // we are building the test environment in ./spec-bundle.js
    // files: [ { pattern: '*.js', watched: false } ],
    files: [
      'lib/*.js',
      'spec/*.js'
    ],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: [
      'ChromeHeadless'
      // 'PhantomJS'
    ],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};


lib/fatigue_check.js
function fatigue_check(event, occurrences, interval, keepalive_occurrences, keepalive_interval) {

    // my defaults
    var occurrences = occurrences || 1; // only the first occurrence
    var interval = interval || 1800; // and every 30 minutes thereafter
    var allow_resolution = true;  // allow resolution events through
    var suppress_flapping = true; // suppress when flapping
    var default_check_interval = 60; // use when check interval cannot be determined from history with cron scheduler

    // set keepalive defaults to match initially, if not provided
    var keepalive_occurrences = keepalive_occurrences || occurrences;
    var keepalive_interval = keepalive_interval || interval;

    // Use the above variable name with the check annotations fatigue_check/
    // (e.g. fatigue_check/occurrences) to override the defaults above

    // Check annotations first
    try {
        if (event.check.hasOwnProperty("annotations")) {
            if (event.check.annotations.hasOwnProperty("fatigue_check/occurrences")) {
                occurrences = parseInt(event.check.annotations["fatigue_check/occurrences"], 10);
            }
            if (event.check.annotations.hasOwnProperty("fatigue_check/interval")) {
                interval = parseInt(event.check.annotations["fatigue_check/interval"], 10);
            }
            if (event.check.annotations.hasOwnProperty("fatigue_check/allow_resolution")) {
                // anything other than explicitly false == true
                allow_resolution = !(/false/i).test(event.check.annotations["fatigue_check/allow_resolution"]);
            }
            if (event.check.annotations.hasOwnProperty("fatigue_check/suppress_flapping")) {
                // anything other than explicitly false == true
                suppress_flapping = !(/false/i).test(event.check.annotations["fatigue_check/suppress_flapping"]);
            }
        }
    }

    catch(err) {
        console.log("failed to get check annotations:");
        console.log(err.message);
        return false;
    }

    // Entity annotations second, to take precedence over check annotations
    // keepalive overrides only exist here
    try {
        if (event.entity.hasOwnProperty("annotations")) {
            if (event.entity.annotations.hasOwnProperty("fatigue_check/occurrences")) {
                occurrences = parseInt(event.entity.annotations["fatigue_check/occurrences"], 10);
            }
            if (event.entity.annotations.hasOwnProperty("fatigue_check/interval")) {
                interval = parseInt(event.entity.annotations["fatigue_check/interval"], 10);
            }
            if (event.entity.annotations.hasOwnProperty("fatigue_check/keepalive_occurrences")) {
                keepalive_occurrences = parseInt(event.entity.annotations["fatigue_check/keepalive_occurrences"], 10);
            }
            if (event.entity.annotations.hasOwnProperty("fatigue_check/keepalive_interval")) {
                keepalive_interval = parseInt(event.entity.annotations["fatigue_check/keepalive_interval"], 10);
            }
            if (event.entity.annotations.hasOwnProperty("fatigue_check/allow_resolution")) {
                // anything other than explicitly false == true
                allow_resolution = !(/false/i).test(event.entity.annotations["fatigue_check/allow_resolution"]);
            }
            if (event.entity.annotations.hasOwnProperty("fatigue_check/suppress_flapping")) {
                // anything other than explicitly false == true
                suppress_flapping = !(/false/i).test(event.entity.annotations["fatigue_check/suppress_flapping"]);
            }
        }
    }

    catch(err) {
        console.log("failed to get entity annotations:");
        console.log(err.message);
        return false;
    }

    if ((/flapping/i).test(event.check.state) && (/true/i).test(suppress_flapping)) {
        return false;
    }
    if (event.is_resolution && (/true/i).test(allow_resolution)) {
        // Check the event.occurrences_watermark to see if we would
        // have allowed the event into the pipeline. This is to prevent
        // resolution event into the pipeline if we've not previously allowed
        // through based on the desired occurrences
        if (event.check.occurrences_watermark >= occurrences) {
            return true;
        }
        else {
            return false;
        }
    } else if (event.is_resolution) {  // allow_resolution must be false, don't allow
        return false;
    }
    if (event.check.name == 'keepalive' && event.check.occurrences == keepalive_occurrences) {
        return true;
    }
    if (event.check.name != 'keepalive' && event.check.occurrences === occurrences) {
        return true;
    }
    if (event.check.interval > 0) {
        // Simple enough, using interval scheduler
        checkInterval = event.check.interval;
    } else if (event.check.interval === 0 && event.check.cron.length > 0 && (!!event.check.history) && event.check.history.length > 1) {
        // If using the cron scheduler and we have sufficient history, compute the check interval based on history
        checkInterval = event.timestamp - event.check.history[event.check.history.length - 2].executed;
    } else {
	// The edge case of cron scheduler and insufficient history, just assume the default
        checkInterval = default_check_interval;
    }

    // The Math.ceil rounds up in the event that the interval requested is not
    // multiple of the check interval
    // Keeaplives
    if (event.check.name == 'keepalive' && event.check.occurrences > keepalive_occurrences && (event.check.occurrences % (Math.ceil(keepalive_interval / checkInterval))) === 0) {
        return true;
    }
    // All other checks
    if (event.check.name != 'keepalive' && event.check.occurrences > occurrences && (event.check.occurrences % (Math.ceil(interval / checkInterval))) === 0) {
        return true;
    }

    return false;
}
